[5주차] 고양이 장난감가게 유효성 추가, 회원 관리 만들기

TaeWoongKim2 opened the pull request:
#### Follow the lecture "유효성 검사"

- Validation
  - Java validation: `@NotBlank`, `@NotNull`.
  - Lombok: `@Builder`, `@NoArgsConstructor`, `@AllArgsConstructor`.

- Object Mapper
  - Dozer for object mapper (source -> targer).

TaeWoongKim2 committed:
Follow the lecture "유효성 검사"

- Validation
  - Java validation: `@NotBlank`, `@NotNull`.
  - Lombok: `@Builder`, `@NoArgsConstructor`, `@AllArgsConstructor`.

- Object Mapper
  - Dozer for object mapper (source -> targer).

TaeWoongKim2 commented:
### 과제 목표

1. 고양이 장난감 가게 유효성 검사 추가하기

    - [x]  고양이 장난감의 필수 입력 항목 - 이름, 메이커, 가격
    - [x]  고양이 장난감 목록 얻기 - GET /products
    - [x]  고양이 장난감 상세 조회하기 - GET /products/{id}
    - [x]  고양이 장난감 등록하기 - POST /products
    - [x]  고양이 장난감 수정하기 - PATCH /products/{id}
    - [x]  고양이 장난감 삭제하기 - DELETE /products/{id}

2. 회원 관리하기

    - [x]  회원의 필수 입력 항목 - 이름, 이메일, 비밀번호
    - [x]  회원 생성하기 - POST /users
    - [x]  회원 수정하기 - PATCH /users/{id}
    - [x]  회원 삭제하기 - DELETE /users/{id}


### 요구 사항

    1. 테스트 커버리지 100%를 달성해야 합니다.
    2. 모든 API 테스트를 통과해야 합니다.
    3. 모든 E2E 테스트를 통과해야 합니다.

TaeWoongKim2 committed:
Modified the Controller Advisor

- Change the advisor name to "ExceptionAdvisor"

- Add new exception handlers
  - MethodArgumentNotValidException
    - for when some of request parameters are invalid.

  - HttpMessageNotReadableException
    - for when required request body is missing.

TaeWoongKim2 committed:
Modified the package configuration

- web
  - controller
  - dto
  - common(exception, helper method)

- core
  - JPA(domain, repository)
  - Service(business logic)

TaeWoongKim2 commented:
제가 이전부터 좋아하는 패키지 구성으로 변경해봤습니다.

강의에서 Web의 DTO와 JPA의 Domain이 서로 다른 관심사를 가지고 있다고 말씀해주셨습니다.

최대한 서로 분리하여 개발하는 것이 추후 변경이나 유지보수에 있어서 서로 영향을 덜 미칠 것으로 생각됩니다.

이러한 이유로 패키지 구성을 변경해봤는데, 위의 말씀에  잘 대응되지 않나 생각이 듭니다 : )

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +51,9 @@ dependencies {
     compileOnly 'org.projectlombok:lombok:1.18.16'
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    // DozerMapper
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
코드 리뷰의 형식으로 DozerMapper의 도입 배경과 근거에 대해 설명해 보세요. 좋은 훈련이 될 거에요.

TaeWoongKim2 replied:
### Dozer란 무엇인가?

Dozer는 한 객체 필드의 데이터를 다른 객체로 재귀적으로 복사해주는 Mapper이다. 그래서, "Dozer Mapper"라고 불리며, Java Bean to Java Bean 기능을 지원한다. Dozer는 양방향 매핑 지원하는데, 다양한 데이터 타입를 가지고 있는 객체 A에서 객체 B로 매핑이 대부분 가능하다. Collection 속성까지 되기 때문에 문서를 잘 읽어보고 활용해볼 수 있을 듯 하다.

### Dozer Mapper의 특징

- 속성 이름 간 매핑을 지원한다.
- 속성 이름뿐만 아니라 유형 간 자동 형변환도 지원한다.
- 양방향 매핑을 지원한다.
- 데이터베이스의 내부 도메인 객체가 외부 프리젠테이션 레이어나 외부 소비자로 번지지 않도록 도와준다.
   또한, 도메인 객체를 외부 API 호출에 매핑하거나 그 반대로 매핑 할 수 있다. (양방향 매핑 지원)
- 리플렉션을 통해 대부분의 필드 매핑을 자동으로 수행하며,
   XML 혹은 @Mapping 어노테이션을 통해 사용자 정의 매핑을 설정 할 수 있다. ( 리플렉션이란? )

 
### 왜, Dozer를  도입했을까?

객체 별로 관심사를 분리가 필요하여 도입하게 되었다. 예를 들어, REST API 서버를 만든다고 가정해보자. 사용자의 요청 처리를 하는 Controller가 존재할 것이다. Contoller는 사용자의 데이터를 받는 객체 A라는 엔티티가 있을 것이다. 보통은 이 객체는 DB Domain과 그대로 카멜기법으로 1:1 매칭시켜놨을 것이다. 과거를 회상해보자.

요청 파라미터를 추가하거나 다른 데이터 가공을 위해 DB Domain 뿐만 아니라 xxxFlag, xxxDatetime 등등 다양하게 추가했던 과거를... 나중에 보면 뭐가 DB 도메인이고, 뭐가 추가된 부분인지 구분하기 어려울 뿐더러 변경할 때마다, Controller - Service - domain 까지 전부 영향을 받게 된다. 분명히 관심사 영역이 다름에도 이렇게 사용하고 있었다. 관심사대로 분리하기 위해서 WEB Request Data(DTO)와 DB Domain Entity로 나누고, 상호 간에 쉽게 매핑할 수 있도록 Dozer를 도입하게 되었다. 또한, Dozer는 변환시 자동 형변환을 지원하기 때문에 Number에서 String으로 변경돼도, Dozer Mapper가 자동으로 해결해주므로 애플리케이션에 별 다른 타격 없이 계속 작동할 수 있다. 굳!
 
Dozer를 통해 Controller (DTO A) ↔ Service (Entity A') ↔ ORM(JPA) ↔ DB 의 흐름으로 만들 수 있게 된 것이다. 이러면, 사용자에게 주는 응답 데이터 항목에 실제 DB 테이블 컬럼명을 노출시키지 않아도 된다는 장점도 존재한다.

 
### 왜, Dozer 여야 했을까?

> 그러게, 왜 Dozer를 사용한걸까?

이미 여러 가지 Object Mapper에 대해 고수들은 성능 측정을 마친 상태이다. Dozer는 속도가 느린 편에 속한다고 대부분의 결과에서 보여진다. 그럼에도 Dozer는 @애노테이션을 통해 코드를 깔끔하게 만들 수 있다는 점이 도입의 가장 큰 이유이다.

 

Getter/Setter를 통하여 수동으로 작업해주는게 가장 빠른 방법이라고 명시되어 있다.
하지만, 이 방법을 한 번 해보면 안다. 코드가 길어진다. 눈이 피곤해진다.

TaeWoongKim2 replied:
#### 참고 자료

**○ About Dozer**
- http://dozer.sourceforge.net/documentation/about.html

**○ Object Mapper 성능 비교**
- https://www.baeldung.com/java-performance-mapping-frameworks


**※ 모르던 용어, "리플렉션"**
- https://dublin-java.tistory.com/53

> Java Object Mapper 가 상당히 많은데 여태 찾아볼 생각도 못했었네요 ㅎㅎ...

johngrib replied:
리플렉션은 어느 언어가 되었건 개발자라면 반드시 한번쯤은 열심히 연습해 보아야 하는 영역입니다. 그런데 중요한 것이 있는데요, 만약 만들려는 것이 프레임워크가 아니라면 실제 회사에서는 거의 절대로 혼자만의 판단으로 리플렉션을 사용하지 않는 것이 좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/ProductService.java:
```diff
@@ -12,9 +13,13 @@
 @Service
 @Transactional
 public class ProductService {
```
`@Service`나 `@Controller` 같은 주요 클래스에는 JavaDoc 주석을 작성해 주세요. 물론 모든 public 메소드에도 JavaDoc 주석을 작성해 주세요.

TaeWoongKim2 replied:
넵, 주석 작성하겠습니다 : )

johngrib commented on app/src/main/java/com/codesoom/assignment/web/dto/ErrorResponse.java:
```diff
@@ -0,0 +1,14 @@
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
이 두 애노테이션을 사용할 때 어떤 점들을 주의해야 하나요?

TaeWoongKim2 replied:
#### `@Builder` 주의사항

아래처럼 `@Builder`와 `@NoArgsConstructor`를 함께 사용하면, 컴파일 시 에러가 발생한다.

```java
@Builder
@NoArgsConstructor
public class Product {
    private String name;
    private Integer price
}
```

`@Builder`를 사용할 때는 모든 필드를 포함하는 생성자가 필요하다. 그래서, 강의에서 나왔다시피 `@Builder` 과 `@NoArgsConstructor`, `@AllArgsConstructor` 세 개의 애노테이션이 쌍으로 사용된다고 언급된 듯하다.

> Finally, applying `@Builder` to a class is as if you added `@AllArgsConstructor(access = AccessLevel.PACKAGE)` to the class and applied the `@Builder` annotation to this all-args-constructor. (아래의 출처 내용 중 일부)

**[출처]**
https://projectlombok.org/features/Builder


#### `@NoArgsConstructor` 주의사항

파라미터가 없는 생성자를 생성해주는 애노테이션이다. 사용에 주의할 점은 다음과 같다.

- `final`로 선언된 필드가 있는 경우, 초기화 할 수 없기 때문에 에러가 발생한다.
  단, `@NoArgsConstructor(force=true)` 옵션을 통해 final 필드를 강제로 초기화 시킬 수 있다.

- `@NonNull` 같이 제약조건이 설정된 필드의 경우, 추후 초기화 전까지 생성자 내 null-check 로직이 생성되지 않기 때문에 사용시 주의가 필요하다.


#### `@AllArgsConstructor` 주의사항

클래스에 존재하는 초기화 되지 않은 final 필드, `@NonNull`과 같이 제약 조건이 설정되어 있는 필드, 이 외의 필드에 대한 생성자를 자동으로 생성한다.  사용에 주의할 점은 다음과 같다.

- _예) 두 개의 동일 타입 인스턴스 멤버를 선언한 상황에서 개발자가 해당 멤버의 순서를 바꾼다면 어떻게 될까?_
   개발자가 인식하지 못하는 사이에 lombok이 생성자의 파라미터 순서를 필드 선언 순서에 따라 변경한다. 이때, 두 필드가 동일 타입이기 때문에 기존 소스에서 오류가 발생하지 않아 아무런 문제없이 동작하는 경우가 발생할 수 있다. 하지만, 실제로 입력된 값이 바뀌어 들어가는 상황이 발생할 수 있기 때문에 사용에 주의가 필요하다. 
(이 부분은 실제로 그러한지 테스트가 필요할 것 같다...)


**[출처]**
https://velog.io/@rosa/Lombok-%EC%A7%80%EC%96%91%ED%95%B4%EC%95%BC-%ED%95%A0-annotation

johngrib reviewed (commented):
`ProductContollerTest`에 보면 다음과 같이 deprecated 된 상수를 사용하는 코드가 있습니다. 해당 상수의 JavaDoc을 읽고 JavaDoc이 시키는대로 조치해 보세요.

![image](https://user-images.githubusercontent.com/1855714/120326716-fc5a9a80-c323-11eb-9ab8-f203acbb8339.png)

TaeWoongKim2 committed:
Make the member controller, dto, and domain.

- Make a MemberController.
  - POST /user for register new member.
  - DELETE /user/:id for delete registed member.
  - PATCH /user/:id for update registed member.

- Make a DTO named "MemberData" for web.

- Make a domain named "Member" for JPA.
  - Check the member table creatation.

TaeWoongKim2 committed:
Make a JPA Member Repository

- Make the interface named "MemberRepository".
- Make the interface named "JpaMemberRepository" extends "MemberRepository" and "CrudRepository".

TaeWoongKim2 committed:
컨트롤러, 서비스, 도메인 관심사 연결 작업.

- MemberConroller - MemberService - MemberRepository - Member

TaeWoongKim2 committed:
Refactoring the MemberController.

- Change `@RequestMapping` "/user" to "/users".
- Add a new exception handler named "MemberNotFoundException" in Advisor.

TaeWoongKim2 committed:
Refectoring the MemberData validation.

- Add a `@Validated` for groups.
  - Make a MemberDataValidation class
    - interface create
    - interface update

  - Change `@NotBlank`
    - to `@NotBlank(groups = {MemberDataValidation.create.class, MemberDataValidation.update.class})`
    - to `@NotBlank(groups = {MemberDataValidation.create.class})`

TaeWoongKim2 committed:
MemberData에 유효성 검사 리팩토링.

- 회원 생성과 수정에 대한 유효성 검사 상이.
  - `@Validated` 활용.

  - MemberDataValidation 클래스 생성.
    - 회원 생성 유효성 그룹 interface create.
    - 회원 수정 유효성 그룹 interface update.

  - `@NotBlank`에 그룹 속성 부여.
    - `@NotBlank(groups = {MemberDataValidation.create.class})`
    - `@NotBlank(groups = {MemberDataValidation.create.class, MemberDataValidation.update.class})`

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+
+    // TODO 1. 회원 생성하기
+
+    /**
+     * 신규 회원 정보를 저장합니다.
+     * @param memberData
+     * @return 저장한 회원
+     */
+    public Member saveMember(MemberData memberData) {
```
메소드 JavaDoc에서 가장 신경써야 하는 것은 메소드 리턴값입니다. 이 설명을 어떻게 수정해야 좋을까요?

다음은 `String.class`의 스크린샷입니다.
![image](https://user-images.githubusercontent.com/1855714/120480672-4443f500-c3ea-11eb-8ece-6af3acc1953e.png)

TaeWoongKim2 replied:
```java

/**          
 * 신규 회원 정보를 저장합니다.
 * @param MemberData -  신규 등록을 요청할 회원 정보.
 * @return Member object - 신규 등록된 회원 정보 반환.
 * /
```

이렇게 바꾸면 어떨까요?

johngrib replied:
`Returns`에 주목해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+        this.memberRepository = memberRepository;
+    }
+
+    // TODO 1. 회원 생성하기
```
이 TODO 는 달성했나요?

TaeWoongKim2 replied:
달성했으나, 주석을 지우지 않았군요!! 삭제하겠습니다 : )

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // TODO 2. 회원 삭제하기
+    public void deleteMember(Long id) {
```
`TODO`로 작성하지 마시고 `public` 메소드에는 JavaDoc으로 설명을 작성해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/domain/Member.java:
```diff
@@ -0,0 +1,33 @@
+@Entity
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
```
이 두 애노테이션을 사용할 때 주의해야 할 점이 있습니다. 조사해서 공유해 주세요.

TaeWoongKim2 replied:
위에 조사했습니다 : )
이외의 이유도 있는지 다른 수각생들 글도 참고해야겠습니다 ㅎㅎ..

johngrib commented on app/src/main/java/com/codesoom/assignment/web/controller/ExceptionAdvisor.java:
```diff
@@ -0,0 +1,66 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice
+public class ExceptionAdvisor {
```
모든 예외를 이 클래스 한 곳에서 모두 처리하나요?

TaeWoongKim2 replied:
음.. 생각하지 못한 질문이십니다!!
이 어플리케이션을 기준으로 만들다 보니, 모든 예외를 한 곳에서 처리해도 되겠다 싶어서 위처럼 만들었습니다.

만약, 더 큰 서비스라면 분리가 필수일 듯 합니다!

johngrib replied:
작을때부터 분리하는 연습을 해보세요. `작으니까 괜찮겠지`랑 `지금은 이렇게 하는 것이 최선이다`는 다릅니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/web/controller/MemberController.java:
```diff
@@ -0,0 +1,66 @@
+
+import javax.validation.Valid;
+
+@RestController
+@RequestMapping("/users")
+public class MemberController {
```
이 클래스를 설명하는 JavaDoc 주석을 작성해 주세요. 이 클래스의 책임은 무엇인가요?

johngrib commented on app/src/main/java/com/codesoom/assignment/web/exception/MemberNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+package com.codesoom.assignment.web.exception;
+
+public class MemberNotFoundException extends RuntimeException {
```
Java의 유명한 표준 예외들에 어떤 형식의 JavaDoc 주석이 달려있는지 조사해 보시고, 이 클래스에 JavaDoc 주석을 작성해 보세요.

어떤 예외를 먼저 봐야 할 지 모르겠다면 `IndexOutOfBoundsException` 부터 봐보세요.

TaeWoongKim2 replied:
Exception은 어떻게 주석을 달아야할지 몰랐는데, 가까운 곳에 참고할 것들이 있었군요!! 꿀팁 감사합니다 : )

johngrib replied:
예외는 던지는 것이라는 걸 기억해두세요!

TaeWoongKim2 replied:
요새 새로운 것을 시도해보고, 또 좋은 습관을 배운다는 느낌이 드네요.
항상 리뷰 감사합니다 : )

johngrib commented on app/src/main/java/com/codesoom/assignment/web/exception/MemberNotFoundException.java:
```diff
@@ -0,0 +1,11 @@
+package com.codesoom.assignment.web.exception;
+
+public class MemberNotFoundException extends RuntimeException {
+    /**
+     * ID에 해당하는 회원을 찾을 수 없음.
+     * @param id
+     */
```
주석이 말하는 바와 코드의 표현이 일치하고 있을까요?

TaeWoongKim2 replied:
위에 말씀해주신 부분 참고하여 수정하겠습니다 : )
위와 같이 쓰면서도 이게 맞나 싶었는데, 질문에 뜨끔했습니다 허허허...

TaeWoongKim2 committed:
ProductController에 JavaDoc 생성.

- Comment, param, and return.

TaeWoongKim2 committed:
ProductService에 JavaDoc 추가.

- 설명, 파라미터, 리턴값.

TaeWoongKim2 committed:
설명, 파라미터, 리턴값, 예외 등을 작성했습니다.

- MemberController.
- MemberService.
- Member domain (DTO, Entity).

TaeWoongKim2 committed:
Add `@Transactional` in MemberService.

for MemberService.updateMember()

- findById and then update that member.

TaeWoongKim2 committed:
공통 예외처리, Product 관련 예외처리, Member 관련 예외처리를 핸들링하도록 분리.

- CommonExceptionAdvisor
  - request body is missing에 대한 예외 핸들링.
  - request data is not valid에 대한 예외 핸들링.

- ProductExceptionAdvisor
  - 특정 고양이장난감을 찾을 수 없을 경우에 대한 예외 핸들링.

- MemberExceptionAdvisor
  - 특정 회원을 찾을 수 없을 경우에 대한 예외 핸들링.

TaeWoongKim2 committed:
Exception 에 대한 JavaDoc 작성.

- `IndexOutOfBoundsException`를 참고하여 JavaDoc 작성.

TaeWoongKim2 committed:
Add a `@CrossOrigin` for 'Access-Control-Allow-Origin' header.

- Applied in `ProductController`.
- Applied in `MemberController`.

TaeWoongKim2 committed:
Make the tests.

- ProductTest & MemberTest
  - Test for Lombok(`@Builder`, `@NoArgsConstructor`, `@AllArgsConstructor`).
  - Test for method that changeWith().

TaeWoongKim2 committed:
Add a controller filter for encoding UTF-8.

- set `.addFilters(new CharacterEncodingFilter("UTF-8", true))`.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/MemberService.java:
```diff
@@ -0,0 +1,68 @@
+        this.memberRepository = memberRepository;
+    }
+
+    /**
+     * 신규 회원를 등록합니다.
+     * @param  memberData - 신규 등록할 회원 정보(이름, 비밀번호, 이메일 등).
+     * @return member - 고유 ID를 발급받은 신규 회원.
+     */
```
스탠다드 라이브러리의 JavaDoc 주석들이 `Returns`로 가득했던 것을 떠올려 보세요. 리턴값에 대한 설명이 가장 중요합니다. 처음 보는 매우 복잡한 코드가 있다고 가정해 봅시다. 그때 "무엇을 넣으면 무엇이 나온다"는 주석이 위에 있다면 엄청 반가울 거에요. 그 주석이 사실이라면 그만큼 도움되는 주석도 없을 겁니다.

```suggestion
    /**
     * 신규 회원을 등록하고, 등록된 회원을 리턴합니다.
     * @param  memberData 신규 등록할 회원 정보
     * @return member 저장된 신규 회원
     */
```

`@param`에 달린 설명은 필요 이상으로 자세했어요. 이름, 비밀번호, 이메일은 지금은 `MemberData`에 소속된 필수정보지만 만약 보안 이슈가 생겨서 비밀번호를 다른 클래스로 분리하게 됐다고 생각해 봅시다. 그렇다면 이 클래스의 이 메소드까지 일부러 찾아와서 저 `비밀번호`라는 주석을 삭제해 주는 귀찮은 일을 해야겠죠. 주석과 비밀번호 필드 사이에 커플링이 생겨 버린 것입니다. 주석에도 객체지향 원칙이 적용된다는 점을 꼭 기억해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/MemberService.java:
```diff
@@ -0,0 +1,68 @@
+        return memberRepository.save(member);
+    }
+
+    /**
+     * 기존 회원를 삭제합니다.
+     * @param  id - 삭제할 특정 회원의 고유 ID.
+     * @throws MemberNotFoundException - ID에 해당하는 회원이 없다면, 회원 정보를 찾을 수 없다는 에러 발생.
+     */
```
`기존`이라는 단어는 아예 안 쓰는 것이 좋습니다. 이 메소드를 호출하는 시점에서 `기존`회원이 아닌 회원도 있을까요? 있다 하더라도 `기존`회원인지 `기존`회원이 아닌지가 이 메소드에서 중요할까요? 

```suggestion
    /**
     * 회원 정보를 찾아 삭제합니다.
     * @param  id 삭제할 회원의 식별자
     * @throws MemberNotFoundException 회원을 찾을 수 없는 경우
     */
```

`@throws`는 어떠한 경우인지를 설명해주는 것으로 충분합니다. `~한다면 회원 정보를 찾을 수 없다는 에러 발생`은 필요 이상으로 설명하고 있어요.

TaeWoongKim2 replied:
당연한 것을 기입할 필요는 없겠네요 ㅎㅎㅎ..

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/MemberService.java:
```diff
@@ -0,0 +1,68 @@
+        memberRepository.delete(member);
+    }
+
+    /**
+     * 기존 회원 정보를 갱합니다.
+     * @param id - 갱신할 특정 회원의 고유 ID.
+     * @param memberData - 갱신할 회원의 정보(이름, 비밀번호, 이메일 등).
+     * @return member - 갱신한 회원 정보 반환.
+     * @throws MemberNotFoundException - ID에 해당하는 회원이 없다면, 회원 정보를 찾을 수 없다는 에러 발생.
+     */
```
여기는 직접 수정해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/core/application/ProductService.java:
```diff
@@ -0,0 +1,91 @@
+import javax.transaction.Transactional;
+import java.util.List;
+
+/**
+ * 고양이 장난감 데이터를 가공하여 반환하거나 처리합니다.
+ */
+@Service
+@Transactional
+public class ProductService {
```
주석에서는 `고양이 장난감`을 말하고 있는데 클래스 이름은 `ProductService`에요. 어째서 클래스 이름을 `CatToyService`라고 짓지 않았나요? 아마도 클래스 이름은 적당히 추상화된 이름을 써야 한다고 생각하셨을 거에요. 지금은 고양이 장난감만 팔고 있지만 2달 후에는 고양이 간식도 팔고 있을 수 있고, 3달 후에는 강아지 사료도 팔고 있을 수 있고, 4달 후에는 고양이와 강아지가 그려진 판촉물을 팔고 있을 수도 있습니다. 

그런 순간이 와도 클래스 이름이 `ProductService`이기 때문에 클래스 이름은 바꾸지 않아도 될 거에요.

하지만 주석은 수정해야겠죠?
이 주석을 수정하시는 것이 좋을까요? 아니면 그냥 놔두시는 것이 좋을까요?

TaeWoongKim2 replied:
오.. 클래스 이름이 추상적이기 때문에 주석도 추상적으로 하면 좋겠군요.
상품이 추가되어도 주석도 변경할 필요가 없고, 추가된 상품의 것들만 어댑터처럼 끼면 되겠네요!!
좋은 조언 감사합니다 : )

johngrib commented on app/src/main/java/com/codesoom/assignment/core/domain/Product.java:
```diff
@@ -9,6 +9,9 @@
 import javax.persistence.GeneratedValue;
 import javax.persistence.Id;
 
+/**
+ * 고양이 장난감 - DB 도메인
+ */
```
이 클래스가 표현하려는 것은 무엇인가요? 이 주석에 `DB`가 노출된 것이 살짝 아쉽습니다.

TaeWoongKim2 replied:
제품이며, DB 도메인을 나타낸다는 것입니다.
`제품 도메인`으로만 해도 충분할 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/web/dto/MemberData.java:
```diff
@@ -0,0 +1,33 @@
+import javax.validation.constraints.NotBlank;
+
+/**
+ * 고양이 장난감 데이터 - Web DTO
```
주석과 코드 내용이 맞지 않는 것 같아요.

TaeWoongKim2 replied:
`제품 DTO`만 해도 충분할 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/web/dto/MemberDataValidation.java:
```diff
@@ -0,0 +1,6 @@
+public class MemberDataValidation {
+    public interface create {}
+    public interface update {}
+}
```
파일 마지막에 개행 문자를 하나 추가해 주세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/web/exception/ProductNotFoundException.java:
```diff
@@ -0,0 +1,15 @@
+package com.codesoom.assignment.web.exception;
+
+/**
+ * 고양이 장난감 ID가 유효하지 않거나 찾을 수 없음을 나타냅니다.
+ */
```
클래스 이름과 주석의 괴리를 좁혀 주세요.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요.

TaeWoongKim2 committed:
Add a new line in MemberDataValidation

- For GitHub.

TaeWoongKim2 committed:
Modified JavaDoc in MemberService

- 더 명확하고, 더 읽기 쉽게 JavaDoc 수정.
  - 불필요하게 자세한 설명 삭제.
  - '기존'이라는 중요하지 않은 설명 삭제.
  - `@throws`는 어떠한 경우에 발생하는지 설명.

TaeWoongKim2 committed:
Product 서비스에 대한 JavaDoc 수정.

- Product라는 추상적인 클래스명과 동일하게 JavaDoc 수정.
  - '고양이 장난감' -> '제품'
  - JavaDoc 주석도 추상화하여 적는 것이 나중에 제품의 종류가
    추가되더라도 유연하게 대처가 가능하다!

TaeWoongKim2 committed:
`ProductNotFoundException`에 대한 JavaDoc 수정.

- 클래스 명칭과 매칭되는 주석으로 수정.

TaeWoongKim2 committed:
Product 도메인 및 DTO JavaDoc 수정.

- Product Domain
  - `DB`라는 단어가 노출되는 것을 수정.
  - `제품`이라는 단어로 대체.
- Product DTO
  - `DTO`라고 간단하게 명시.
  - `제품`이라는 단어로 대체.
